### PR TITLE
Moj/otomi chart

### PR DESCRIPTION
**Pull request by mojtabaimani**
Adding a task to map a plain values.yaml from otomi chart to different value files and merge with current bootstrapped ones.

### CODE REVIEW

File: src/tasks/otomi/otomi-chart.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+      bsValues = yaml.load(fs.readFileSync(bsPath).toString())
+    }
+    if (!bsValues) {
+      bsValues = {}
+    }
+
+    merge(bsValues, valueObject)
+    fs.writeFileSync(bsPath, yaml.safeDump(bsValues))
+  } catch (e) {
+    console.log(e)
```
Comment (j-zimnowoda) [inline src/tasks/otomi/otomi-chart.ts:44]: We should allow this task to fail if an exception occurs

File: src/tasks/otomi/otomi-chart.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+      const valueObject = {
+        charts: {
+          [chart]: plainValues.charts[chart],
+        },
+      }
+      mergeValues(chart, valueObject, `${env.OTOMI_VALUES_TARGET}/charts`)
+    })
+
+    console.log('Done.')
+  } catch (e) {
```
Comment (j-zimnowoda) [inline src/tasks/otomi/otomi-chart.ts:91]: Why do you catch it?

File: src/tasks/otomi/otomi-chart.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+const env = cleanEnv({
+  OTOMI_VALUES_INPUT,
+  OTOMI_SCHEMA_PATH,
+  OTOMI_VALUES_TARGET,
+})
+
+const schemaKeywords = ['properties', 'anyOf', 'allOf', 'oneOf']
+
+function extractSecrets(schema: any, parentKey: string): any {
```
Comment (j-zimnowoda) [inline src/tasks/otomi/otomi-chart.ts:16]: needs  unit tests

File: src/tasks/otomi/otomi-chart.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+      if (typeof childObj !== 'object') return false
+      if ('x-secret' in childObj) return `${parentKey}.${key}`
+      if (key in ['anyOf', 'allOf', 'oneOf']) return extractSecrets(childObj, parentKey)
+      const address = schemaKeywords.includes(key) ? parentKey : `${parentKey}.${key}`
+      return extractSecrets(childObj, address)
+    })
+    .filter(Boolean)
+}
+
+function mergeValues(cat: string, valueObject, folder: string): void {
```
Comment (j-zimnowoda) [inline src/tasks/otomi/otomi-chart.ts:29]: needs unit tests

**Review by j-zimnowoda — CHANGES_REQUESTED** [review]
Nice work. Needs some unit tests.

File: src/tasks/otomi/otomi-chart.test.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        },
+      },
+    }
+    const expectedResult = [
+      'root.pullSecret',
+      'root.provider.aws.password',
+      'root.provider.google.password',
+      'root.provider.azure.password',
+    ]
+    const x = extractSecrets(schema, 'root')
```
Comment (Morriz) [inline src/tasks/otomi/otomi-chart.test.ts:36]: so leaving 'root' empty or undefined even would create '.pullSecret'? That could be another test, but I am not sure we need that. I am curious though

File: src/tasks/otomi/otomi-chart.ts [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+async function main(): Promise<void> {
+  const values = yaml.safeLoad(fs.readFileSync(env.OTOMI_VALUES_INPUT, 'utf8')) as any
+
+  // creating secret files
+  const schema = yaml.safeLoad(fs.readFileSync(env.OTOMI_SCHEMA_PATH, 'utf8')) as any
+  const derefSchema = await $RefParser.dereference(schema)
+  const cleanSchema = omit(derefSchema, ['definitions', 'properties.teamConfig']) // FIXME: lets fix the team part later
+  const secretsJsonPath = extractSecrets(cleanSchema, 'root').map((str) => str.replace('root.', ''))
```
Comment (Morriz) [inline src/tasks/otomi/otomi-chart.ts:48]: why the need to provide 'root'? Just write the function to make that param optional...

**Review by Morriz — CHANGES_REQUESTED** [review]
looking good...minor things

File: src/tasks/otomi/otomi-chart.ts [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+      charts: {
+        [chart]: plainValues.charts[chart],
+      },
+    }
+    mergeValues(`${env.OTOMI_VALUES_TARGET}/charts/${chart}.yaml`, valueObject)
+  })
+
+  console.log('otomi chart values merged with the bootstrapped values.')
+}
+main()
```
Comment (j-zimnowoda) [inline src/tasks/otomi/otomi-chart.ts:86]: if (typeof require !== 'undefined' && require.main === module) {
  main()
}

So when you import this module the function is not executed

File: src/tasks/otomi/otomi-chart.test.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+      additionalProperties: false,
+      definitions: {
+        rawValues: { description: 'bla bla', type: 'object' },
+      },
+      properties: {
+        apps: { description: 'console.', type: 'object' },
+        pullSecret: {
+          description: 'console.',
+          type: 'string',
+          [`x-secret`]: 'true',
```
Comment (j-zimnowoda) [inline src/tasks/otomi/otomi-chart.test.ts:17]: 'x-secret': 'true',

File: src/tasks/otomi/otomi-chart.ts [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    bsValues = yaml.load(fs.readFileSync(targetPath).toString())
+  }
+  if (!bsValues) {
+    bsValues = {}
+  }
+  merge(bsValues, valueObject)
+  fs.writeFileSync(targetPath, yaml.safeDump(bsValues))
+}
+
+async function main(): Promise<void> {
```
Comment (j-zimnowoda) [inline src/tasks/otomi/otomi-chart.ts:43]: export default

File: src/tasks/otomi/otomi-chart.ts [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+const env = cleanEnv({
+  OTOMI_VALUES_INPUT,
+  OTOMI_SCHEMA_PATH,
+  OTOMI_VALUES_TARGET,
+})
+
+const schemaKeywords = ['properties', 'anyOf', 'allOf', 'oneOf']
+
+export default function extractSecrets(schema: any, parentAddress?: string): Array<string> {
```
Comment (j-zimnowoda) [inline src/tasks/otomi/otomi-chart.ts:16]: just export

**Review by j-zimnowoda — CHANGES_REQUESTED** [review]
Impressive work. Few minor remarks

**Review by j-zimnowoda — APPROVED** [review]

**Comment by Morriz** [comment]
@j-zimnowoda why do you approve a non working solution?

**Comment by Morriz** [comment]
And maybe we should start a PR in "draft" mode so it can't be approved @mojtabaimani

**Comment by j-zimnowoda** [comment]
> @j-zimnowoda why do you approve a non working solution?

I saw it working and it passed my code review. Draft mode it is not a solution. We still need to invest in integration test framework, because it is not possible to test each PR manually.

**Review by Morriz — APPROVED** [review]

**Pull request merged**